Frontend-unit-tests-guide.md

```diff
@@ -1,409 +1,484 @@
+* [Contacts](#contacts)
+
+## Introduction
+
+### Goal
+
+Our goal is for every file with frontend code to be thoroughly tested by its associated test file. For example, `user.service.ts` should be comprehensively tested by `user.service.spec.ts`. By "comprehensively," we mean that the tests should cover all possible ways the code in `user.service.ts` might be used. It's especially important for test to cover both happy (expected) and unhappy (unexpected) usages. [Here](https://github.com/oppia/oppia/blob/ae649aa08f/core/templates/services/questions-list.service.spec.ts#L19) is a good example of a comprehensive set of test cases.
+
+### Unit tests
+
+A unit test checks the behavior of small pieces of code, which are called units. Often, the unit is a funtion, in which case the behavior being tested would be the function's outputs given some inputs.
+
+A unit test should depend only on the "external" behavior to be tested, not the specific implementation of the function. This means that the behavior of other units is out of scope.
 [Here](https://github.com/oppia/oppia/blob/ae649aa08f1375457ec9e3c90257197b68fec7cd/core/templates/domain/skill/RubricObjectFactorySpec.ts#L28-L41) is a simple unit test that demonstrates these points.
 
+### Code coverage
+
+We get a rough idea of how comprehensive our tests are by measuring _code coverage_, which is the fraction of testable lines of code that were executed when the tests were run. For example, consider the following pseudocode:
+
+```text
+1 # Compute the absolute value of a number
+2 function absoluteValue(number) {
+3   if number <= 0 {
+4     return -number
+5   } else {
+6     return number
+7   }
+8 }
+```
+
+Now consider the following sets of test cases:
+
+* `absoluteValue(1)` and `absoluteValue(5)`: These test cases are not comprehensive because they only test positive numbers. Code coverage is also not 100% because line 4 is never executed.
+* `absoluteValue(0)` and `absoluteValue(1)`: These test cases are not comprehensive because they do not test negative numbers, and it's important for an absolute value function to correctly handle negative inputs. However, the code coverage is 100% because both blocks of the `if` statement are executed.
+* `absoluteValue(-1)`, `absoluteValue(0)`, and `absoluteValue(1)`: These test cases are comprehensive, and code coverage is 100%. Note that even though line 1 doesn't execute, coverage is 100% because line 1 is not executable.
+
+This example illustrates something very important about code coverage: **Code coverage less than 100% implies that the tests are not comprehensive, but code coverage of 100% does not imply that tests are comprehensive.** Therefore, while code coverage is a useful tool, you should primarily think about whether your tests cover all the possible behaviors of the code being tested--you should have a behavior-first perspective. Don't just think about which lines are covered.
```

Comment:
Perhaps replace "--" with "In other words,"; at the moment it's not entirely clear what "--" implies and I had to read it twice.

Also, possibly the part to bold should be "whether your tests cover all the possible behaviors of the code being tested". That seems like the key thing to emphasize here.

Reply:
done